### PR TITLE
Prevent autosave when the post is locked

### LINKED ISSUE #21236: The block editor does not trigger the interim login modal early enough
**Describe the bug**

Given a user who gets logged out while they're editing a post, the block editor does not always trigger the interim login modal before the user sees an `Updating failed` error message.

**To reproduce**
1. Create a new post in the block editor
2. In a separate tab, log out of the site (or clear your browser cookies for the domain)
3. Click `Save Draft`
4. Observe that the `Updating failed` message appears
5. Do nothing and wait a while for autosave to kick in, and observe that the interim login modal does appear, but too late

**Expected behavior**

When a user action triggers a server side request (for example saving the post) but the user's session is no longer valid, they should always be presented with the interim login modal instead of the `Updating failed` error message.

**Screenshots**

Screenshot demonstrating that the interm modal does appear, but only after the user has seen the `Updating failed` error message.

### LINKED ISSUE #36118: Mixed http/https access to admin can cause infinite loop of nonce checks
### Description

Gutenberg can find itself in an infinite loop of REST requests due to an invalid nonce if a user uses their site over https but at some point happens to load up the login screen over http. This happened to me in a local environment where I don't force https traffic but have a cert available. I inadvertently navigated to the http address, realized my mistake and re-navigated to the https address, and when I created a post in Gutenberg it was unusable.

### Step-by-step reproduction instructions

1. Sign into wp-admin successfully at https address, e.g. `https://wp.test/wp-admin/`.
2. Load up the `http://` variant of wp-admin, e.g. `http://wp.test/wp-admin/`. You will get the login screen but don't sign in.
3. Load up wp-admin again at the https address, e.g. `https://wp.test/wp-admin/`
4. Open browser console and go to network tab
5. Navigate to create a new post in GB, e.g. `https://wp.test/wp-admin/post-new.php`
6. Observe an infinite loop of 403 requests in the network tab of the browser console due to invalid nonces.

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/465154/139695037-6ed80b0f-3361-44fa-98a8-063dbe33fe91.mp4

### Environment info

- WordPress 5.8.1 using Twenty Twenty-One theme, no plugins active
- Chrome 95.0.4638.54
- MacOS 11.6

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### LINKED ISSUE #42400: More conservative number of retry attempts for failed requests
### Description

While testing to see if [Core-13779](https://core.trac.wordpress.org/ticket/13779) was still happening in WordPress 6.0, I found that it was, and actually causes potentially greater problems in the block editor.

TLDR of that ticket: When the WordPress files are in a directory and a site is set up to run as such (site.com/wp) and the site is changed to instead be visible from the site's root (site.com), all users who are logged in will not be properly authenticated. This is because the cookie set on login does not exist for the new location.

In the past, this caused issues with the Classic Editor and previewing posts. However, in WordPress 6.0, if you open the block editor, there are repeated requests to the REST API that return 403 responses. After opening the editor with my browser console open, ~500 requests were made in 30 seconds.

Regardless of the root cause of the 403 responses, the editor should not continually retry the same requests infinitely.

### Step-by-step reproduction instructions

1. Create a WordPress install in a subdirectory (`/wp` for example).
2. Log in, and [follow steps 2-9 found here](https://wordpress.org/support/article/giving-wordpress-its-own-directory/#moving-process).
3. Open the browser's console and click the network tab.
4. Create a new post or edit an old one.
5. Observe the repeated requests.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by desaiuditd**
## Description
Fixes #13509. Potentially addresses #21236, #42400, #36118 as well. Also fixes a bug described below.

## Steps to reproduce
- Login as User A.
- Create a post in block editor. Let's say with Post ID - 1. Do not publish it.
- Make any changes right before the user session is about to expire and wait for session expiry and do not save the post.
  - Testing Note: One way to achieve this is by making some changes in the post content and immediately deleting all the cookies. Do not save the post.
- Wait for the on-page WordPress login popup to appear. Stay on the screen. Do not log back in just yet.
- Login as User B in different browser or incognito mode.
- Wait for the post lock to expire for User A on Post with ID - 1.
- Open the Post with ID - 1 as User B once post is not locked anymore.
- Observe that the post has last persisted changes made by User A.
- Make some changes in the post as User B and save it. Do not publish it.
- Leave this window open for User B.
- Meanwhile, go back to User A window and re-login.
- Notice from dev console that an autosave request is triggered with User A's last unsaved changes from the open browser.
- Also notice that until next heartbeat happens, you can actually go ahead and make changes and save the post as User A after re-login, even though the post is locked by User B.
- Come back to User B window and refresh the page.
- Notice that you lost changes made by User B.

## How has this been tested?
With the fix in this PR, I tested above reproduction steps. After User A re-logs in, there's not autosave request triggered, User A immediately sees the `PostLockedModal` preventing any further updates and User B is not loosing the content anymore.

## Types of changes
- Non-breaking change
- Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by desaiuditd** [comment]
Some of the changes I've made here ... they can be questionable. Would like to get any initial feedback and iterate on this, if needed. Please feel free to point me to right direction if I've missed or unknowingly neglected anything.

File: packages/editor/src/store/selectors.js [hunk 1]
```diff
@@ -1128,7 +1132,7 @@ export function getPermalinkParts( state ) {
  * @return {boolean} Is locked.
  */
 export function isPostLocked( state ) {
-	return state.postLock.isLocked;
+	return state.postLock?.isLocked ?? false;
```
Comment (desaiuditd) [inline packages/editor/src/store/selectors.js:1038]: ℹ️  This is irrelevant change to the issue. Without this certain unit tests started to fail complaining about postLock being undefined in the state. This will just make sure that it has valid fallback.

File: packages/editor/src/components/post-locked-modal/index.js [hunk 2]
```diff
@@ -89,7 +152,12 @@ export default function PostLockedModal() {
 			const received = data[ 'wp-refresh-post-lock' ];
 			if ( received.lock_error ) {
 				// Auto save and display the takeover modal.
-				autosave();
+				// Meanwhile, other user may have opened the post and started to make new changes.
+				// Post may have been published or may have been switched back to draft.
+				// In either case, we don't want to accidentally loose the changes made by other user.
+				// If we perform server side autosave, and the post is in draft, previous changes will be overwritten in the database.
+				// Hence, we will only perform local autosave.
+				autosave( { local: true } );
```
Comment (desaiuditd) [inline packages/editor/src/components/post-locked-modal/index.js:163]: ℹ️  Ideally, `isPostLocked` check (which is part of `isEditedPostSaveable` check which in turn is part of `isEditedPostAutosaveable`) should be able to prevent this server autosave request. But somehow it seems like the data store at point of time in code execution is not properly updated. Hence, I'm still observing the network request for the autosave in the browser. Hence, I decided to put `local: true` to prevent the server autosave.

File: packages/editor/src/components/post-locked-modal/index.js [hunk 3]
```diff
@@ -44,18 +50,75 @@ export default function PostLockedModal() {
...
+			} else if (
+				! isCurrentUserSessionActive &&
+				heartbeatTickTriggered &&
+				data[ 'wp-auth-check' ]
+			) {
+				// If current login state is false and server responds with auth check for previously triggered heartbeat,
+				// update login state to true.
+				updateCurrentUserSession( { isActive: true } );
+			}
+		}
```
Comment (desaiuditd) [inline packages/editor/src/components/post-locked-modal/index.js:116]: ℹ️  This whole function (which is managing the current user session in data store) can be questionable and there could be better way to handle this. So I'm happy to take any advise of improvement here.

File: packages/editor/src/components/post-locked-modal/index.js [hunk 4]
```diff
@@ -23,17 +23,23 @@ import PostPreviewButton from '../post-preview-button';
...
+	const {
+		isDraft,
 		isLocked,
 		isTakeover,
 		user,
 		postId,
 		postLockUtils,
 		activePostLock,
 		postType,
+		isCurrentUserSessionActive,
```
Comment (desaiuditd) [inline packages/editor/src/components/post-locked-modal/index.js:44]: ℹ️  For now, this data (`isCurrentUserSessionActive`) is only being used in this component and could well be a local state instead of being in the data store. But I can think of few use cases where we would want to check for current user sessions. I myself thought about using the same in `AutosaveMonitor` component. (This was when I was exploring potential fix in this PR. Eventually the need did not arise. So I removed that code before committing.) But the point is, there is a use case and can be a need in future.

**Comment by georgeh** [comment]
Thanks for tackling this. I was able to reproduce the error in trunk, but was not able to get your branch working in my wp-env environment. Could you rebase the branch and I'll try again?

This does look like a good approach. I initially was thinking that this should be rejecting the updates on the server side, but it looks like that is already happening and the client is working around that rejection.

**Comment by desaiuditd** [comment]
@georgeh Sorry, I've not been able to get back to this PR. I got busy with other things. I'll try to revive this PR this weekend.

**Comment by desaiuditd** [comment]
@georgeh I think this is ready for review.

**Comment by georgeh** [comment]
@desaiuditd sorry I haven't been keeping up with the project and am not in a position to re-review

**Comment by desaiuditd** [comment]
@desrosj @Mamaduka @mrfoxtalbot @johnbillion @aristath @ellatrix @ntsekouras @nerrad @jsnajdr @tyxla @andrewserong 

How can I move this PR forward? Can any of you help to review and test it?

**Comment by desaiuditd** [comment]
I've tested and verified on my local after syncing my branch with latest `trunk`. I can still see the bug occurring on `trunk` and this branch fixes it.

**Comment by desaiuditd** [comment]
@WordPress/gutenberg-core @WordPress/gutenberg Apologies for a wider ping here. But this PR has got stale without any attention for quite some time now.

Can I get some feedback on the code changes and help to review, test and move it forward? Don't want it to become stale again.

**Comment by jsnajdr** [comment]
Hi @desaiuditd :wave: I'm looking at this, trying to understand how do the heartbeats and post locks work.

One thing that immediately comes to mind are the `admin-ajax.php` requests. It would be nice if we had a REST endpoint for this.

File: packages/editor/src/components/post-preview-button/index.js [hunk 5]
```diff
@@ -106,8 +106,8 @@ export default function PostPreviewButton( {
 	role,
 	onPreview,
 } ) {
-	const { postId, currentPostLink, previewLink, isSaveable, isViewable } =
-		useSelect( ( select ) => {
+	const { postId, currentPostLink, previewLink, isViewable } = useSelect(
```
Comment (swissspidy) [inline packages/editor/src/components/post-preview-button/index.js:109]: Why remove the `isSaveable` check in this component?

`isEditedPostSaveable` returns `false` if the post is considered empty and it doesn't make sense to preview an empty post.

File: packages/editor/src/components/post-preview-button/test/index.js [hunk 6]
```diff
@@ -139,22 +139,6 @@ describe( 'PostPreviewButton', () => {
 		).toBeInTheDocument();
 	} );
 
-	it( 'should be disabled if post is not saveable.', () => {
```
Comment (swissspidy) [inline packages/editor/src/components/post-preview-button/test/index.js]: As per my other comment, it doesn't seem right to remove this functionality.

File: packages/editor/src/store/selectors.js (hunk 1 above)
Comment (swissspidy) [reply packages/editor/src/store/selectors.js:1038]: I think in the test file we usually just made sure a default is set. So for consistency probably best to update those tests too.

**Comment by swissspidy** [comment]
> One thing that immediately comes to mind are the `admin-ajax.php` requests. It would be nice if we had a REST endpoint for this.

I once built a custom REST endpoint for post locking so that I don't have to use the heartbeat API:

https://github.com/GoogleForCreators/web-stories-wp/blob/0634e91314aeffb5fa461d84240f0eb54f9efd3f/includes/REST_API/Stories_Lock_Controller.php

File: packages/editor/src/components/post-locked-modal/index.js (hunk 2 above)
Comment (swissspidy) [reply packages/editor/src/components/post-locked-modal/index.js:163]: > If we perform server side autosave, and the post is in draft, previous changes will be overwritten in the database.

I don't think that's true. It creates a _new_ autosave, not overriding any of the other user's changes. The original post and the other user's autosave both still exist.

So this change would break that.

FWIW it's possible to write e2e tests around post locking and autosaves... Sounds like now would be a good time to do that :-)

File: packages/editor/src/components/post-locked-modal/index.js [hunk 7]
```diff
@@ -141,7 +211,7 @@ export default function PostLockedModal() {
 			removeAction( 'heartbeat.tick', hookName );
 			window.removeEventListener( 'beforeunload', releasePostLock );
 		};
-	}, [] );
+	}, [ isDraft, isLocked ] );
```
Comment (swissspidy) [inline packages/editor/src/components/post-locked-modal/index.js:214]: Is this intentional? More dependencies means the effect is triggered more often. I think an empty dependencies array was deliberately chosen here.

File: packages/core-data/src/selectors.ts [hunk 8]
```diff
@@ -185,6 +186,21 @@ export function getCurrentUser( state: State ): ET.User< 'edit' > {
 	return state.currentUser;
 }
 
+/**
+ *Returns a boolean flag for current user's active login session.
```
Comment (swissspidy) [inline packages/core-data/src/selectors.ts:190]:
```suggestion
 * Returns a boolean flag for current user's active login session.
```

File: packages/editor/src/components/post-locked-modal/index.js [hunk 9]
```diff
@@ -54,9 +67,61 @@ export default function PostLockedModal() {
...
+		 */
+		function checkUserLogin( data ) {
+			// Bail early, if the heartbeat has neither auth check data attribute nor nonce expiry attribute.
+			if (
+				! [ 'wp-auth-check', 'nonces_expired' ].some( ( attr ) =>
+					Object.prototype.hasOwnProperty.call( data, attr )
+				)
+			) {
+				return;
+			}
```
Comment (swissspidy) [inline packages/editor/src/components/post-locked-modal/index.js:91]: `auth-check.js` in core only looks at `wp-auth-check`, not `nonces_expired`, see https://github.com/WordPress/wordpress-develop/blob/7b24083859c37d8ac75b9549f458bc389610eaef/src/js/_enqueues/lib/auth-check.js#L162-L168

So I'd remove the `nonces_expired` check here.

And then simplify to:

```suggestion
			if ( ! 'wp-auth-check' in data ) {
				return;
			}
```

**Comment by Mamaduka** [comment]
@spacedmonkey proposed porting that or a similar API - #37789.

File: packages/core-data/src/actions.js [hunk 10]
```diff
@@ -57,6 +57,20 @@ export function receiveCurrentUser( currentUser ) {
...
 }
 
+/**
+ * Returns an action object used to lock the editor.
+ *
+ * @param {Object} currentUserSession Meta information about the login session for current user.
+ *
+ * @return {Object} Action object.
+ */
+export function __experimentalUpdateCurrentUserSession( currentUserSession ) {
```
Comment (youknowriad) [inline packages/core-data/src/actions.js:67]: This looks like local state. core-data was not really mean to store things like that. So can you explain more the reasoning here and what this is about?

File: packages/editor/src/components/post-locked-modal/index.js [hunk 11]
```diff
@@ -54,9 +67,61 @@ export default function PostLockedModal() {
...
+				heartbeatTickTriggered &&
+				data[ 'wp-auth-check' ]
+			) {
+				// If current login state is false and server responds with auth check for previously triggered heartbeat,
+				// update login state to true.
+				updateCurrentUserSession( { isActive: true } );
+			}
+		}
+
+		addAction( 'heartbeat.tick', checkUserLoginHookName, checkUserLogin );
```
Comment (youknowriad) [inline packages/editor/src/components/post-locked-modal/index.js:118]: We already subscribe to this action in the hook down here. Was it a deliberate choice to separate these two logics. The reasoning is not clear to me.

File: packages/core-data/src/actions.js (hunk 10 above)
Comment (jsnajdr) [reply packages/core-data/src/actions.js:67]: I think this state could be very easily moved to the `editor` store, to the `state.postLock` object. That object already has plenty of fields (`isLocked`, `isTakeover`, `user`, ...), so why not add another `isUserActive` field there?

All other aspects of this user session check should be integrated into the post lock state, too. Why add another hook to `heartbeat.tick`, if we can just expand the existing `receivePostLock` function?

Comment (swissspidy) [reply packages/core-data/src/actions.js:67]: Moving this to the `postLock` object sounds confusing. Post lock state is coming from an entirely different API response and it's easy to accidentally override `postLock.isUserActive` (or vice-versa) in the process.

Comment (jsnajdr) [reply packages/core-data/src/actions.js:67]: Then it could be renamed to `state.heartbeat`, with subfields for post lock, auth check and possibly other things. They are all requested together, and also used together.

File: packages/editor/src/components/post-locked-modal/index.js (hunk 9 above)
Comment (jsnajdr) [reply packages/editor/src/components/post-locked-modal/index.js:91]: If we used `apiFetch` for the heartbeat request, then we wouldn't need to care about nonces at all, because `apiFetch` handles them internally (nonce middleware). Automatically requesting a new one when the old one expires etc.

I'm not sure if `admin-ajax.php` is compatible with `apiFetch` or whether we'd need to create a REST endpoint first.

Comment (swissspidy) [reply packages/editor/src/components/post-locked-modal/index.js:91]: No that would require a REST endpoint. See the ither comments about that.

This comment here is still relevant for admin-ajax

**Review by SirLouen — CHANGES_REQUESTED** [review]
Apart from the reviews already made, this patch is currently not applying cleanly anymore.

**Comment by SirLouen** [comment]
Feel free to unblock this. I cannot unblock it myself other than accepting changes.
